{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "1Xe75WARCWi_",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c589f12b-4d2e-4daa-962f-b8bdfb5c56dd"
      },
      "source": [
        "# import os\n",
        "# from google.colab import drive\n",
        "# drive.mount('/content/drive', force_remount=True)\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yXkBX34qDfCh",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5cd2160f-ac02-4b71-eaa3-b0dcb171fd2b"
      },
      "source": [
        "# !pip install import-ipynb\n",
        "# import import_ipynb"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting import-ipynb\n",
            "  Downloading import_ipynb-0.1.4-py3-none-any.whl (4.1 kB)\n",
            "Requirement already satisfied: IPython in /usr/local/lib/python3.7/dist-packages (from import-ipynb) (5.5.0)\n",
            "Requirement already satisfied: nbformat in /usr/local/lib/python3.7/dist-packages (from import-ipynb) (5.4.0)\n",
            "Requirement already satisfied: pygments in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (2.6.1)\n",
            "Requirement already satisfied: traitlets>=4.2 in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (5.1.1)\n",
            "Requirement already satisfied: prompt-toolkit<2.0.0,>=1.0.4 in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (1.0.18)\n",
            "Requirement already satisfied: simplegeneric>0.8 in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (0.8.1)\n",
            "Requirement already satisfied: pickleshare in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (0.7.5)\n",
            "Requirement already satisfied: pexpect in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (4.8.0)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (57.4.0)\n",
            "Requirement already satisfied: decorator in /usr/local/lib/python3.7/dist-packages (from IPython->import-ipynb) (4.4.2)\n",
            "Requirement already satisfied: wcwidth in /usr/local/lib/python3.7/dist-packages (from prompt-toolkit<2.0.0,>=1.0.4->IPython->import-ipynb) (0.2.5)\n",
            "Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python3.7/dist-packages (from prompt-toolkit<2.0.0,>=1.0.4->IPython->import-ipynb) (1.15.0)\n",
            "Requirement already satisfied: jupyter-core in /usr/local/lib/python3.7/dist-packages (from nbformat->import-ipynb) (4.10.0)\n",
            "Requirement already satisfied: jsonschema>=2.6 in /usr/local/lib/python3.7/dist-packages (from nbformat->import-ipynb) (4.3.3)\n",
            "Requirement already satisfied: fastjsonschema in /usr/local/lib/python3.7/dist-packages (from nbformat->import-ipynb) (2.15.3)\n",
            "Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import-ipynb) (21.4.0)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import-ipynb) (4.11.4)\n",
            "Requirement already satisfied: importlib-resources>=1.4.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import-ipynb) (5.7.1)\n",
            "Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import-ipynb) (0.18.1)\n",
            "Requirement already satisfied: typing-extensions in /usr/local/lib/python3.7/dist-packages (from jsonschema>=2.6->nbformat->import-ipynb) (4.1.1)\n",
            "Requirement already satisfied: zipp>=3.1.0 in /usr/local/lib/python3.7/dist-packages (from importlib-resources>=1.4.0->jsonschema>=2.6->nbformat->import-ipynb) (3.8.0)\n",
            "Requirement already satisfied: ptyprocess>=0.5 in /usr/local/lib/python3.7/dist-packages (from pexpect->IPython->import-ipynb) (0.7.0)\n",
            "Installing collected packages: import-ipynb\n",
            "Successfully installed import-ipynb-0.1.4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ie3UxrfdDfeH",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e079644d-6a3f-48a0-f07a-672c9b431a8e"
      },
      "source": [
        "# %cd \"drive/MyDrive/Bionformatics_Project/Colab\""
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/.shortcut-targets-by-id/1he9S1Es-XalZ9vgCRMbTTNd6cXeRkkFI/Bionformatics_Project/Colab\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SiFtJil5EnnD"
      },
      "source": [
        "if the above step are not commented we get an error when calling this notebook from another one, because it cannot find a way to ask the authorization code.\n",
        "Anyway given the fact that one time I runned the code below with the drive mounted at the current directory we have no need to run it again because the file are in drive, and we can access them if we have drive mounted in the main script."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5j4Pc9st81l-",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4920b753-fae4-4284-ae0b-4a1c0cc6e6fb"
      },
      "source": [
        "# #Train/Valid/Test set for predicting median gene expression levels in the human\n",
        "# !wget -r -np -nH --reject \"index.html*\" --cut-dirs 6 \\\n",
        "#  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest/\n",
        "# #Train/Valid/Test set for predicting median gene expression levels in the mouse\n",
        "# !wget -r -np -nH --reject \"index.html*\" --cut-dirs 6 \\\n",
        "#  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/\n",
        "# #Prepare set of input sequences to generate predictions\n",
        "# !wget https://xpresso.gs.washington.edu/data/Xpresso-predict.zip\n",
        "# !unzip Xpresso-predict.zip"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/\n",
            "Resolving krishna.gs.washington.edu (krishna.gs.washington.edu)... 128.208.8.136\n",
            "Connecting to krishna.gs.washington.edu (krishna.gs.washington.edu)|128.208.8.136|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:50 (72.8 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html.tmp’ saved [1415/1415]\n",
            "\n",
            "Loading robots.txt; please ignore errors.\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/robots.txt\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 404 Not Found\n",
            "2022-06-21 07:43:50 ERROR 404: Not Found.\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=N;O=D\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=N;O=D.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:50 (77.0 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=N;O=D.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=N;O=D.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=M;O=A\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=M;O=A.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:50 (73.9 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=M;O=A.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=M;O=A.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=S;O=A\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=S;O=A.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:50 (79.6 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=S;O=A.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=S;O=A.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=D;O=A\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=D;O=A.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:50 (77.0 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=D;O=A.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=D;O=A.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:50--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/test.h5\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 16101376 (15M)\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/test.h5’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]  15.36M  50.4MB/s    in 0.3s    \n",
            "\n",
            "2022-06-21 07:43:51 (50.4 MB/s) - ‘pM10Kb_1KTest_Mouse/test.h5’ saved [16101376/16101376]\n",
            "\n",
            "--2022-06-21 07:43:51--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/train.h5\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 318166951 (303M)\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/train.h5’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>] 303.43M  76.2MB/s    in 4.2s    \n",
            "\n",
            "2022-06-21 07:43:55 (72.9 MB/s) - ‘pM10Kb_1KTest_Mouse/train.h5’ saved [318166951/318166951]\n",
            "\n",
            "--2022-06-21 07:43:56--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/valid.h5\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 16098395 (15M)\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/valid.h5’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]  15.35M  58.7MB/s    in 0.3s    \n",
            "\n",
            "2022-06-21 07:43:56 (58.7 MB/s) - ‘pM10Kb_1KTest_Mouse/valid.h5’ saved [16098395/16098395]\n",
            "\n",
            "--2022-06-21 07:43:56--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=N;O=A\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=N;O=A.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:56 (70.6 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=N;O=A.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=N;O=A.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:56--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=M;O=D\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=M;O=D.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:56 (73.2 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=M;O=D.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=M;O=D.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:56--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=S;O=D\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=S;O=D.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:56 (72.0 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=S;O=D.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=S;O=D.tmp since it should be rejected.\n",
            "\n",
            "--2022-06-21 07:43:56--  https://krishna.gs.washington.edu/content/members/vagar/Xpresso/data/datasets/pM10Kb_1KTest_Mouse/?C=D;O=D\n",
            "Reusing existing connection to krishna.gs.washington.edu:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1415 (1.4K) [text/html]\n",
            "Saving to: ‘pM10Kb_1KTest_Mouse/index.html?C=D;O=D.tmp’\n",
            "\n",
            "pM10Kb_1KTest_Mouse 100%[===================>]   1.38K  --.-KB/s    in 0s      \n",
            "\n",
            "2022-06-21 07:43:56 (76.9 MB/s) - ‘pM10Kb_1KTest_Mouse/index.html?C=D;O=D.tmp’ saved [1415/1415]\n",
            "\n",
            "Removing pM10Kb_1KTest_Mouse/index.html?C=D;O=D.tmp since it should be rejected.\n",
            "\n",
            "FINISHED --2022-06-21 07:43:56--\n",
            "Total wall clock time: 6.5s\n",
            "Downloaded: 12 files, 334M in 4.7s (70.7 MB/s)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6uXufw4u9I1X"
      },
      "source": [
        "import sys, gzip, h5py, pickle, os\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib.colors import ListedColormap\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import warnings\n",
        "warnings.filterwarnings(\"ignore\")\n",
        "import csv\n",
        "import seaborn as sns\n",
        "import random\n",
        "import json\n",
        "\n",
        "random_seed = 43\n",
        "random.seed(random_seed)\n",
        "\n",
        "import sys\n",
        "import numpy\n",
        "numpy.set_printoptions(threshold=100)\n",
        "pd.set_option('display.max_columns', None)\n",
        "pd.set_option('display.width', 1000)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-dklwR9BAsZP"
      },
      "source": [
        "datadir = {'human':\"Dataset/pM10Kb_1KTest\", 'mouse':'Dataset/pM10Kb_1KTest_Mouse'} # the one used by Xpresso\n",
        "\n",
        "class DataManager():\n",
        "    def __init__(self, \n",
        "                 datadir         = datadir, \n",
        "                 transformer     = False, \n",
        "                 micro           = False, \n",
        "                 tf              = False, \n",
        "                 datadir_micro   = \"Dataset/microRNA FINALE\", \n",
        "                 datadir_tf      = \"Dataset/dataset_aumentati\", \n",
        "                 remove_indicted = False,\n",
        "                 DeepLncLoc      = False,\n",
        "                 kmer            = 3,\n",
        "                 species         = 'human' # human mouse\n",
        "                 ):\n",
        "        \n",
        "        self.datadir         = datadir[species]\n",
        "        self.remove_indicted = remove_indicted\n",
        "        self.indicted        = [577, 1494, 2729, 2829, 4095, 5038, 5340, 7804, 8445, 8532, 8557, 11040, 11369, 11638, 11650, 11785, 12216, 13851, 13927]\n",
        "        self.indicted_test   = [437]\n",
        "        self.X_traintf       = [0]\n",
        "        self.X_validtf       = [0]\n",
        "        self.X_testtf        = [0]\n",
        "        self.DeepLncLoc      = DeepLncLoc\n",
        "        self.kmer            = kmer\n",
        "        self.species         = species\n",
        "\n",
        "        if not DeepLncLoc:\n",
        "            #translated versions\n",
        "            if tf:\n",
        "                self.trainfile = h5py.File(os.path.join(datadir_tf, 'train_tf.h5'), 'r')\n",
        "                self.validfile = h5py.File(os.path.join(datadir_tf, 'validation_tf.h5'), 'r')\n",
        "                self.testfile  = h5py.File(os.path.join(datadir_tf, 'test_tf.h5'), 'r')\n",
        "                \n",
        "                self.X_trainhalflife = self.trainfile['halflife'] \n",
        "                self.X_trainpromoter = self.trainfile['promoter']\n",
        "                self.X_traintf       = self.trainfile['tf']\n",
        "                self.y_train         = self.trainfile['label'] \n",
        "                self.geneName_train  = self.trainfile['gene']\n",
        "\n",
        "                self.X_validhalflife = self.validfile['halflife']\n",
        "                self.X_validpromoter = self.validfile['promoter']\n",
        "                self.X_validtf       = self.validfile['tf']\n",
        "                self.y_validid         = self.validfile['label'] \n",
        "                self.geneName_valid  = self.validfile['gene']\n",
        "\n",
        "                self.X_testhalflife  = self.testfile['halflife']\n",
        "                self.X_testpromoter  = self.testfile['promoter']\n",
        "                self.X_testtf        = self.testfile['tf']\n",
        "                self.y_test          = self.testfile['label']\n",
        "                self.geneName_test   = self.testfile['gene']\n",
        "            \n",
        "            elif micro:\n",
        "            \n",
        "                self.trainfile = h5py.File(os.path.join(datadir_micro, 'microRNA_train.h5'), 'r')\n",
        "                self.validfile = h5py.File(os.path.join(datadir_micro, 'microRNA_val.h5'), 'r')\n",
        "                self.testfile  = h5py.File(os.path.join(datadir_micro, 'microRNA_test.h5'), 'r')\n",
        "                \n",
        "                self.X_trainhalflife = self.trainfile['halflife'] \n",
        "                self.X_trainpromoter = self.trainfile['promoter']\n",
        "                self.X_traintf       = self.trainfile['tf']\n",
        "                self.X_trainmicro    = self.trainfile['micro']\n",
        "                self.y_train         = self.trainfile['label'] \n",
        "                self.geneName_train  = self.trainfile['gene']\n",
        "\n",
        "                self.X_validhalflife = self.validfile['halflife']\n",
        "                self.X_validpromoter = self.validfile['promoter']\n",
        "                self.X_validtf       = self.validfile['tf']\n",
        "                self.X_valmicro      = self.validfile['micro']\n",
        "                self.y_validid         = self.validfile['label'] \n",
        "                self.geneName_valid  = self.validfile['gene']\n",
        "\n",
        "                self.X_testhalflife  = self.testfile['halflife']\n",
        "                self.X_testpromoter  = self.testfile['promoter']\n",
        "                self.X_testtf        = self.testfile['tf']\n",
        "                self.X_testmicro     = self.testfile['micro']\n",
        "                self.y_test          = self.testfile['label']\n",
        "                self.geneName_test   = self.testfile['gene']\n",
        "            \n",
        "\n",
        "            else:\n",
        "                self.trainfile = h5py.File(os.path.join(self.datadir, 'train.h5'), 'r')\n",
        "                self.validfile = h5py.File(os.path.join(self.datadir, 'valid.h5'), 'r')\n",
        "                self.testfile  = h5py.File(os.path.join(self.datadir, 'test.h5'),  'r')\n",
        "                \n",
        "                self.X_trainhalflife = self.trainfile['data'] \n",
        "                self.X_trainpromoter = self.trainfile['promoter'] \n",
        "                self.y_train         = self.trainfile['label'] \n",
        "                self.geneName_train  = self.trainfile['geneName']\n",
        "\n",
        "                self.X_validhalflife = self.validfile['data']\n",
        "                self.X_validpromoter = self.validfile['promoter']\n",
        "                self.y_validid         = self.validfile['label'] \n",
        "                self.geneName_valid  = self.validfile['geneName']\n",
        "\n",
        "                self.X_testhalflife  = self.testfile['data']\n",
        "                self.X_testpromoter  = self.testfile['promoter']\n",
        "                self.y_test          = self.testfile['label']\n",
        "                self.geneName_test   = self.testfile['geneName']\n",
        "\n",
        "            if transformer:\n",
        "                if self.species == \"human\":\n",
        "                    transFile = \"translated_transformers.h5\"\n",
        "                    if not tf:\n",
        "                        self.dictionary = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
        "                        transformersdata   = h5py.File(os.path.join(\"Dataset/translated_promoters\", transFile), 'r')\n",
        "                        if self.species == \"human\":\n",
        "                            self.X_trainpromoter_tr = transformersdata[\"train\"]\n",
        "                            self.X_validpromoter_tr = transformersdata[\"valid\"]\n",
        "                            self.X_testpromoter_tr  = transformersdata[\"test\"] \n",
        "                    else:\n",
        "                        self.dictionary = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
        "                        transformersdata   = h5py.File(os.path.join(\"Dataset/dataset_aumentati\", 'translated_transformers_tf.h5'), 'r')\n",
        "                        self.X_trainpromoter_tr = transformersdata[\"train\"]\n",
        "                        self.X_validpromoter_tr = transformersdata[\"valid\"]\n",
        "                        self.X_testpromoter_tr  = transformersdata[\"test\"]    \n",
        "                else:   # mouse\n",
        "                    transFile = \"translated_transformers_mouse.h5\"\n",
        "                    transformersdata  = pd.read_hdf(f\"Dataset/translated_promoters/{transFile}\")\n",
        "                    self.X_trainpromoter_tr = transformersdata[\"train\"].values[0]\n",
        "                    self.X_validpromoter_tr = transformersdata[\"valid\"].values[0]\n",
        "                    self.X_testpromoter_tr  = transformersdata[\"test\"].values[0]                                          \n",
        "\n",
        "        else:\n",
        "            if kmer == 3:\n",
        "              datadir = 'Dataset/embedded_data'\n",
        "            elif kmer == 6:\n",
        "              datadir = 'Dataset/embedded_data/6mer'\n",
        "            elif kmer == 9:\n",
        "              datadir = 'Dataset/embedded_data/9mer'\n",
        "            else:\n",
        "              print('KMER NOT SUPPORTED')\n",
        "\n",
        "            if self.species==\"human\":\n",
        "\n",
        "                if tf:\n",
        "                    trainfile = h5py.File(os.path.join(datadir, 'etrain_tf.h5'), 'r')\n",
        "                    validfile = h5py.File(os.path.join(datadir, 'evalidation_tf.h5'), 'r')\n",
        "                    testfile  = h5py.File(os.path.join(datadir, 'etest_tf.h5'), 'r')\n",
        "\n",
        "                    self.X_trainhalflife, self.X_trainpromoter, self.y_train, self.X_traintf = trainfile['halflife'], trainfile['promoter'], trainfile['label'], trainfile['tf']\n",
        "                    self.X_validhalflife, self.X_validpromoter, self.y_valid, self.X_validtf = validfile['halflife'], validfile['promoter'], validfile['label'], validfile['tf']\n",
        "                    self.X_testhalflife,  self.X_testpromoter,  self.y_test,  self.X_testtf  = testfile['halflife'],  testfile['promoter'],  testfile['label'],  testfile['tf']\n",
        "\n",
        "                    self.X_trainpromoter, self.X_validpromoter, self.X_testpromoter = np.array(self.X_trainpromoter), np.array(self.X_validpromoter), np.array(self.X_testpromoter)\n",
        "                    self.X_trainhalflife, self.X_validhalflife, self.X_testhalflife = np.array(self.X_trainhalflife), np.array(self.X_validhalflife), np.array(self.X_testhalflife)\n",
        "                    self.X_traintf,       self.X_validtf,       self.X_testtf       = np.array(self.X_traintf),       np.array(self.X_validtf),       np.array(self.X_testtf)\n",
        "                    self.y_train,         self.y_valid,         self.y_test         = np.array(self.y_train),         np.array(self.y_valid),         np.array(self.y_test)\n",
        "\n",
        "                else:\n",
        "                    trainfile = h5py.File(os.path.join(datadir, 'etrain.h5'), 'r')\n",
        "                    validfile = h5py.File(os.path.join(datadir, 'evalidation.h5'), 'r')\n",
        "                    testfile  = h5py.File(os.path.join(datadir, 'etest.h5'), 'r')\n",
        "\n",
        "                    self.X_trainhalflife, self.X_trainpromoter, self.y_train = trainfile['halflife'], trainfile['promoter'], trainfile['label']\n",
        "                    self.X_validhalflife, self.X_validpromoter, self.y_valid = validfile['halflife'], validfile['promoter'], validfile['label']\n",
        "                    self.X_testhalflife,  self.X_testpromoter,  self.y_test  = testfile['halflife'],  testfile['promoter'],  testfile['label']\n",
        "\n",
        "                    self.X_trainpromoter, self.X_validpromoter, self.X_testpromoter = np.array(self.X_trainpromoter), np.array(self.X_validpromoter), np.array(self.X_testpromoter)\n",
        "                    self.X_trainhalflife, self.X_validhalflife, self.X_testhalflife = np.array(self.X_trainhalflife), np.array(self.X_validhalflife), np.array(self.X_testhalflife)\n",
        "                    self.y_train,         self.y_valid,         self.y_test         = np.array(self.y_train), np.array(self.y_valid), np.array(self.y_test)\n",
        "\n",
        "            else:\n",
        "                    trainfile = h5py.File(os.path.join(datadir, 'etrain_mouse.h5'), 'r')\n",
        "                    validfile = h5py.File(os.path.join(datadir, 'evalidation_mouse.h5'), 'r')\n",
        "                    testfile  = h5py.File(os.path.join(datadir, 'etest_mouse.h5'), 'r')\n",
        "\n",
        "                    self.X_trainhalflife, self.X_trainpromoter, self.y_train = trainfile['halflife'], trainfile['promoter'], trainfile['label']\n",
        "                    self.X_validhalflife, self.X_validpromoter, self.y_valid = validfile['halflife'], validfile['promoter'], validfile['label']\n",
        "                    self.X_testhalflife,  self.X_testpromoter,  self.y_test  = testfile['halflife'],  testfile['promoter'],  testfile['label']\n",
        "\n",
        "                    self.X_trainpromoter, self.X_validpromoter, self.X_testpromoter = np.array(self.X_trainpromoter), np.array(self.X_validpromoter), np.array(self.X_testpromoter)\n",
        "                    self.X_trainhalflife, self.X_validhalflife, self.X_testhalflife = np.array(self.X_trainhalflife), np.array(self.X_validhalflife), np.array(self.X_testhalflife)\n",
        "                    self.y_train,         self.y_valid,         self.y_test         = np.array(self.y_train), np.array(self.y_valid), np.array(self.y_test)                \n",
        "                \n",
        "\n",
        "\n",
        "\n",
        "        \n",
        "\n",
        "    def get_train(self, np_format=True, translated=False, micro=False):\n",
        "        \n",
        "        if self.DeepLncLoc:\n",
        "            return self.X_trainhalflife, self.X_trainpromoter, self.y_train, self.X_traintf\n",
        "        else:\n",
        "            if np_format:\n",
        "                \n",
        "            \n",
        "                if translated:\n",
        "                    \n",
        "                    if self.remove_indicted:\n",
        "                    \n",
        "                        train_translated_no_indicted_halflife = np.delete(self.X_trainhalflife, self.indicted, 0)\n",
        "                        train_translated_no_indicted_promoter = np.delete(np.array(self.X_trainpromoter_tr), self.indicted, 0)\n",
        "                        train_translated_no_indicted_labels   = np.delete(np.array(self.y_train), self.indicted, 0)\n",
        "                        train_translated_no_indicted_names    = np.delete(np.array(self.geneName_train), self.indicted, 0)\n",
        "                        \n",
        "                        return train_translated_no_indicted_halflife, train_translated_no_indicted_promoter, train_translated_no_indicted_labels, train_translated_no_indicted_names\n",
        "                        \n",
        "                    else:\n",
        "\n",
        "                        train_translated_halflife = np.array(self.X_trainhalflife)\n",
        "                        train_translated_promoter = self.X_trainpromoter_tr\n",
        "                        train_translated_labels   = np.array(self.y_train)\n",
        "                        train_translated_names    = np.array(self.geneName_train)\n",
        "                        \n",
        "                        return train_translated_halflife, train_translated_promoter, train_translated_labels, train_translated_names, np.array(self.X_traintf)\n",
        "\n",
        "                elif micro:\n",
        "\n",
        "                    return np.array(self.X_trainhalflife), np.array(self.X_trainpromoter), np.array(self.y_train), np.array(self.geneName_train), np.array(self.X_traintf), np.array(self.X_trainmicro)\n",
        "\n",
        "                elif self.remove_indicted:\n",
        "\n",
        "                    train_no_indicted_halflife = np.delete(self.X_trainhalflife, self.indicted, 0)\n",
        "                    train_no_indicted_promoter = np.delete(np.array(self.X_trainpromoter), self.indicted)\n",
        "                    train_no_indicted_labels   = np.delete(np.array(self.y_train), self.indicted)\n",
        "                    train_no_indicted_names    = np.array(self.geneName_train)\n",
        "                    \n",
        "                    return train_no_indicted_halflife, train_no_indicted_promoter, train_no_indicted_labels, train_no_indicted_names\n",
        "\n",
        "                else:\n",
        "                    \n",
        "                    return np.array(self.X_trainhalflife), np.array(self.X_trainpromoter), np.array(self.y_train), np.array(self.geneName_train), np.array(self.X_traintf)\n",
        "                \n",
        "            else:\n",
        "                \n",
        "                return self.X_trainhalflife, self.X_trainpromoter, self.y_train, self.geneName_train, self.X_traintf\n",
        "    \n",
        "    def get_validation(self, np_format=True, translated=False, micro=False):\n",
        "\n",
        "        if self.DeepLncLoc:\n",
        "            return self.X_validhalflife, self.X_validpromoter, self.y_valid, self.X_validtf\n",
        "        else:\n",
        "            if np_format:\n",
        "                \n",
        "                if translated:\n",
        "                    \n",
        "                    valid_translated_halflife_np = np.array(self.X_validhalflife)\n",
        "                    valid_translated_promoter_np = self.X_validpromoter_tr\n",
        "                    valid_translated_labels_np   = np.array(self.y_validid)\n",
        "                    valid_translated_names_np    = np.array(self.geneName_valid)\n",
        "                    \n",
        "                    return valid_translated_halflife_np, valid_translated_promoter_np, valid_translated_labels_np, valid_translated_names_np, np.array(self.X_validtf)\n",
        "                \n",
        "                elif micro:\n",
        "\n",
        "                    return np.array(self.X_validhalflife), np.array(self.X_validpromoter), np.array(self.y_validid), np.array(self.geneName_valid), np.array(self.X_validtf), np.array(self.X_valmicro)\n",
        "                \n",
        "                else:\n",
        "                    \n",
        "                    return np.array(self.X_validhalflife), np.array(self.X_validpromoter), np.array(self.y_validid), np.array(self.geneName_valid), np.array(self.X_validtf)\n",
        "                \n",
        "            else:\n",
        "            \n",
        "                return self.X_validhalflife, self.X_validpromoter, self.y_validid, self.geneName_valid, self.X_validtf\n",
        "\n",
        "    def get_test(self, np_format=True, translated=False, micro=False):\n",
        "\n",
        "        if self.DeepLncLoc:\n",
        "            return self.X_testhalflife,  self.X_testpromoter,  self.y_test,  self.X_testtf\n",
        "        else:\n",
        "            if np_format:\n",
        "\n",
        "                if translated:\n",
        "\n",
        "                    if self.remove_indicted:\n",
        "                    \n",
        "                        train_translated_no_indicted_halflife = np.delete(self.X_testhalflife, self.indicted_test, 0)\n",
        "                        train_translated_no_indicted_promoter = np.delete(np.array(self.X_testpromoter_tr), self.indicted_test, 0)\n",
        "                        train_translated_no_indicted_labels   = np.delete(np.array(self.y_test), self.indicted_test, 0)\n",
        "                        train_translated_no_indicted_names    = np.delete(np.array(self.geneName_test), self.indicted_test, 0)\n",
        "                        \n",
        "                        return train_translated_no_indicted_halflife, train_translated_no_indicted_promoter, train_translated_no_indicted_labels, train_translated_no_indicted_names\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        test_translated_halflife_np = np.array(self.X_testhalflife)\n",
        "                        test_translated_promoter_np = self.X_testpromoter_tr\n",
        "                        test_translated_labels_np   = np.array(self.y_test)\n",
        "                        test_translated_names_np    = np.array(self.geneName_test)\n",
        "                    \n",
        "                    return test_translated_halflife_np, test_translated_promoter_np, test_translated_labels_np, test_translated_names_np, np.array(self.X_testtf)\n",
        "\n",
        "                elif micro:\n",
        "                \n",
        "                    return np.array(self.X_testhalflife), np.array(self.X_testpromoter), np.array(self.y_test), np.array(self.geneName_test), np.array(self.X_testtf), np.array(self.X_testmicro)\n",
        "            \n",
        "                else:\n",
        "                \n",
        "                    return np.array(self.X_testhalflife), np.array(self.X_testpromoter), np.array(self.y_test), np.array(self.geneName_test), np.array(self.X_testtf)\n",
        "            \n",
        "            else:\n",
        "                \n",
        "                return self.X_testhalflife, self.X_testpromoter, self.y_test, self.geneName_test, self.X_testtf"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# print(\"> Loading DataManager...\")\n",
        "# dm = DataManager(transformer=True, micro = False, tf = False, species=\"mouse\")\n",
        "\n",
        "# print(\"> Loading train data...\")\n",
        "# X_trainhalflife, X_trainpromoter, y_train, geneName_train, _ = dm.get_train(True, True)\n",
        "\n",
        "# print(\"> Loading validation data...\")\n",
        "# X_validationhalflife, X_validationpromoter, y_validation, geneName_valid, _ = dm.get_validation(True, True)\n",
        "\n",
        "# print(\"> Loading test data...\")\n",
        "# X_testhalflife, X_testpromoter, y_test, geneName_test, _ = dm.get_test(True, True)\n",
        "\n",
        "# X_trainpromoter = np.array(X_trainpromoter)\n",
        "# X_trainpromoter.shape"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Q18Baai5fQEG",
        "outputId": "34f4b88c-f523-47ea-ecf5-0a77b2a061ab"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "> Loading DataManager...\n",
            "> Loading train data...\n",
            "> Loading validation data...\n",
            "> Loading test data...\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([[0, 2, 1, ..., 0, 2, 3],\n",
              "       [1, 0, 0, ..., 1, 1, 1],\n",
              "       [1, 1, 2, ..., 2, 0, 3],\n",
              "       ...,\n",
              "       [2, 2, 0, ..., 3, 0, 3],\n",
              "       [1, 3, 1, ..., 1, 3, 1],\n",
              "       [0, 2, 0, ..., 2, 3, 2]], dtype=uint8)"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# X_testpromoter.shape"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qmLQtDR5gLuF",
        "outputId": "bd35d63d-9cd3-4dc8-eeca-8646a83b7d09"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(1000, 20000)"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    }
  ]
}